File: src/minilake/mcp/tools/secrets.py
"""MCP tools for Secrets (/api/2.0/secrets)."""

from typing import Any

from minilake.mcp.client import MinilakeClient

_PREFIX = "/api/2.0/secrets"


def register(mcp: Any, client: MinilakeClient) -> None:
    @mcp.tool()
    async def create_secret_scope(scope: str) -> dict[str, Any]:
        """Create a secret scope."""
        return await client.post(f"{_PREFIX}/scopes/create", json={"scope": scope})

    @mcp.tool()
    async def list_secret_scopes() -> dict[str, Any]:
        """List all secret scopes."""
        return await client.get(f"{_PREFIX}/scopes/list")

    @mcp.tool()
    async def delete_secret_scope(scope: str) -> dict[str, Any]:
        """Delete a secret scope and its secrets."""
        return await client.post(f"{_PREFIX}/scopes/delete", json={"scope": scope})

    @mcp.tool()
    async def put_secret(scope: str, key: str, string_value: str) -> dict[str, Any]:
        """Store a secret.

        The value can never be read back through the API — that matches real Databricks. It
        becomes readable only inside a job container, by referencing
        `{{secrets/<scope>/<key>}}` in the task's env, which minilake resolves to the real
        value at execution time.
        """
        return await client.post(f"{_PREFIX}/put", json={"scope": scope, "key": key, "string_value": string_value})

    @mcp.tool()
    async def list_secrets(scope: str) -> dict[str, Any]:
        """List the secret keys in a scope (keys and timestamps only, never values)."""
        return await client.get(f"{_PREFIX}/list", params={"scope": scope})

    @mcp.tool()
    async def delete_secret(scope: str, key: str) -> dict[str, Any]:
        """Delete one secret."""
        return await client.post(f"{_PREFIX}/delete", json={"scope": scope, "key": key})

File: ui/src/lib/api/secrets.ts
import { apiFetch } from "./client";
import type { SecretMetadata, SecretScope } from "./types";

const SECRETS = "/api/2.0/secrets";

export async function listScopes(): Promise<SecretScope[]> {
  const data = await apiFetch<{ scopes?: SecretScope[] }>(`${SECRETS}/scopes/list`);
  return data.scopes ?? [];
}

export function createScope(scope: string): Promise<unknown> {
  return apiFetch(`${SECRETS}/scopes/create`, { method: "POST", body: { scope } });
}

export function deleteScope(scope: string): Promise<unknown> {
  return apiFetch(`${SECRETS}/scopes/delete`, { method: "POST", body: { scope } });
}

/**
 * List a scope's keys.
 *
 * Keys and timestamps only — there is deliberately no read path for values.
 * `GET /secrets/get` always fails with BAD_REQUEST, matching real Databricks,
 * where a value is only readable through `dbutils.secrets.get()`.
 */
export async function listSecrets(scope: string): Promise<SecretMetadata[]> {
  const data = await apiFetch<{ secrets?: SecretMetadata[] }>(`${SECRETS}/list`, { query: { scope } });
  return data.secrets ?? [];
}

export function putSecret(scope: string, key: string, value: string): Promise<unknown> {
  return apiFetch(`${SECRETS}/put`, {
    method: "POST",
    body: { scope, key, string_value: value },
  });
}

export function deleteSecret(scope: string, key: string): Promise<unknown> {
  return apiFetch(`${SECRETS}/delete`, { method: "POST", body: { scope, key } });
}
